opm/material/fluidsystems/BlackOilFluidSystem.hpp

```diff
@@ -180,56 +180,57 @@ class BlackOil : public BaseFluidSystem<Scalar, BlackOil<Scalar> >
      */
     static void initFromDeck(const Deck& deck, const EclipseState& eclState)
     {
-        auto densityKeyword = deck.getKeyword("DENSITY");
-        size_t numRegions = densityKeyword.size();
+        size_t numRegions = eclState.runspec().tabdims().getNumPVTTables();
         initBegin(numRegions);
 
         numActivePhases_ = 0;
         std::fill(&phaseIsActive_[0], &phaseIsActive_[numPhases], false);
 
-        if (deck.hasKeyword("OIL")) {
+        const auto& phases = eclState.runspec().phases();
+        if (phases.active(Phase::OIL)) {
             phaseIsActive_[oilPhaseIdx] = true;
             ++ numActivePhases_;
         }
 
-        if (deck.hasKeyword("GAS")) {
+        if (phases.active(Phase::GAS)) {
             phaseIsActive_[gasPhaseIdx] = true;
             ++ numActivePhases_;
         }
 
-        if (deck.hasKeyword("WATER")) {
+        if (phases.active(Phase::WATER)) {
             phaseIsActive_[waterPhaseIdx] = true;
             ++ numActivePhases_;
         }
 
         // this fluidsystem only supports two or three phases
         assert(numActivePhases_ >= 2 && numActivePhases_ <= 3);
 
-        setEnableDissolvedGas(deck.hasKeyword("DISGAS"));
-        setEnableVaporizedOil(deck.hasKeyword("VAPOIL"));
+        setEnableDissolvedGas(eclState.cfg().simulation().hasDISGAS());
```

Comment:
It's a bit clumsy still, but I'm expecting that to change soon™. Half the reason we're doing this is to actually improve our overall interface, making it comfortable to use and that's harder if we also have to expose our intermediate stage.